Apache Milestone Release Prep

matrei opened the pull request.

matrei committed:
build: update build scan and cache settings

matrei committed:
chore: add `HEADER` and `NOTICE` files

matrei committed:
build: update `buildSrc` repositories

matrei committed:
build: update `buildSrc` dependencies

matrei committed:
docs: convert docs from `gdoc` to `asciidoc`

matrei committed:
chore: clean up `test-config.gradle`

matrei committed:
chore: add license headers

matrei committed:
chore: add Apache Release Audit Tool (rat)

matrei committed:
build: add configuration for reproducible builds

matrei committed:
chore: cleanup `docs-config.gradle`

matrei committed:
chore: add verification scripts

matrei committed:
chore: update Gradle wrapper to `8.14.2`

matrei committed:
build: generate redirect file to guide

matrei committed:
ci: update GitHub action workflows

jdaugherty commented on build.gradle:
```diff
@@ -6,6 +29,17 @@ plugins {
+
 repositories {
     mavenCentral()
     maven { url = 'https://repo.grails.org/grails/core' }
```
We need to add a filter on the snapshot repo so we do not get blocked for 404s

jdaugherty commented on buildSrc/build.gradle:
```diff
@@ -9,11 +28,25 @@ file('../gradle.properties').withInputStream {
 
 repositories {
     mavenCentral()
```
Won't we still need the grails restricted repo for spring rc's etc?

matrei replied:
'https://repo.grails.org/grails/restricted' is already added. Is there another "restricted" repo?

jdaugherty commented on gradle/rat-root-config.gradle:
```diff
@@ -0,0 +1,48 @@
+            '**/build/**', // Gradle generated build directories
+            '**/.gitattributes', // git configuration isn't code
+            '**/.gradle/**', '**/wrapper/**', 'gradlew*',  // gradle wrapper files excluded from src zip
+            '**/resources/*', // exclude test artifacts
```
Does this project actually have test resources?

jdaugherty commented on gradle/docs-config.gradle:
```diff
@@ -1,34 +1,104 @@
 
-tasks.withType(PublishGuideTask).configureEach {
-    classpath = configurations.guideConfiguration
+tasks.named('asciidoctor', AsciidoctorTask) {
```
Why move away from the guide task?  We can easily style and have a central theme with that task.

matrei replied:
Haha, collateral damage when converting from `gdoc` to `asciidoc`.

matrei replied:
@jdaugherty On the other hand, The `PublishGuideTask` and `DocPublisher` seems to need some updating and refactoring.
```console
Execution failed for task ':docs'.
> Could not resolve all files for configuration ':docs'.
   > Could not find org.grails:grails-docs:7.0.0-SNAPSHOT.
     Searched in the following locations:
       - https://repo.maven.apache.org/maven2/org/grails/grails-docs/7.0.0-SNAPSHOT/maven-metadata.xml
       - https://repo.maven.apache.org/maven2/org/grails/grails-docs/7.0.0-SNAPSHOT/grails-docs-7.0.0-SNAPSHOT.pom
       - https://repo.grails.org/grails/restricted/org/grails/grails-docs/7.0.0-SNAPSHOT/maven-metadata.xml
       - https://repo.grails.org/grails/restricted/org/grails/grails-docs/7.0.0-SNAPSHOT/grails-docs-7.0.0-SNAPSHOT.pom
```
I'll leave this as is for now.

jdaugherty replied:
These tasks are in grails-gradle.  It's showing the wrong coordinates in your message?

matrei replied:
@jdaugherty I think the coordinates needs updating here:
https://github.com/apache/grails-core/blob/1ed250d77ad922dc6d82e8f214e7d305341e6b6f/grails-gradle/plugins/src/main/groovy/org/grails/gradle/plugin/doc/GrailsDocGradlePlugin.groovy#L44

jdaugherty commented on README.md:
```diff
@@ -1,3 +1,19 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+-->
```
This file needs builds instructions for it to be a valid release.

We also need to add the gradle-bootstrap folder we have added else where

matrei committed:
build(feedback): add repository filters

Prevents getting blocked by requesting Gradle requesting non-existing artifacts in RAO

matrei committed:
build(feedback): remove unnecessary rat exclude

matrei committed:
chore: update `.sdkmanrc`

Bump java version and add gradle version for use in `gradle-bootstrap`.

matrei committed:
build(feedback): add `gradle-bootstrap`

matrei committed:
ci: add `.asf.yaml`

matrei committed:
chore: add `CODE_OF_CONDUCT.md` and `DISCLAIMER`

matrei committed:
build: add a `skipTests` condition

matrei committed:
docs: add `Building from Source` section to `README`

matrei committed:
fix(deps): force Groovy to `4.0.27`

We need `4.0.27` for reproducible builds,
but Spring downgrades to `4.0.26` at the moment.

matrei committed:
fix(deps): bump Grails to `7.0.0-M4`

matrei committed:
ci(feedback): add rat workflow

jamesfredley reviewed (approved):
The only thing we might wait on is the resolution for gdoc.   

In the comment it was trying to pull the snapshot instead of M4 and was not checking the Apache Snapshot repo.

matrei commented:
> The only thing we might wait on is the resolution for gdoc.
> 
> In the comment it was trying to pull the snapshot instead of M4 and was not checking the Apache Snapshot repo.

@jamesfredley I think this needs updating before being able to use the `PublishGuideTask`:
https://github.com/apache/grails-core/blob/1ed250d77ad922dc6d82e8f214e7d305341e6b6f/grails-gradle/plugins/src/main/groovy/org/grails/gradle/plugin/doc/GrailsDocGradlePlugin.groovy#L44

https://github.com/apache/grails-core/pull/14826

jamesfredley commented:
@matrei Since https://github.com/apache/grails-core/blob/v7.0.0-M4/grails-gradle/plugins/src/main/groovy/org/grails/gradle/plugin/doc/GrailsDocGradlePlugin.groovy#L44 was that way on M4, please proceed without it.
